### PR TITLE
mrp attach

### PR DESCRIPTION
**Pull request by lshamis**
# Description

Add `mrp attach` to get a terminal into a running process.
Also, `mrp up <proc> --attach`

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [ ] Refactor
- [ ] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [ ] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

## Before and After

If this PR introduces a change or fixes a bug, please add before and after screenshots (if this is causes a UX change) or before and after examples(this could be plain text, videos etc ) to demonstrate the change.

If this PR introduces a new feature, please also add examples or screenshots demonstrating the feature.

# Testing

Please describe the tests that you ran to verify your changes and how we can reproduce.

# Checklist:

- [ ] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.
- [ ] [Polymetis only] I ran on hardware (1) all scripts in `tests/scripts`, (2) asv benchmarks.

### CODE REVIEW

File: mrp/src/mrp/cmd/up.py [hunk 1]
```diff
@@ -117,6 +120,12 @@ def cli(
     if not names:
         raise ValueError("No processes found")
 
+    if attach and len(names) != 1:
+        raise ValueError(f"Cannot attach to {len(names)} processes")
```
Comment (zephirefaith) [inline mrp/src/mrp/cmd/up.py:124]: Can we add here that attach expects only 1 process as input.

**Comment by zephirefaith** [comment]
Thanks for working on this PR @lshamis and for explaining the code + demoing the utility. This PR adds important functionality which will make debugging, prototyping and testing development around the Alephzero + MRP ecosystem easier and faster.

I added 1 comment to make error message verbose, but apart from that it looks good to me.

**Review by zephirefaith — APPROVED** [review]

File: mrp/src/mrp/cmd/up.py (hunk 1 above)
Comment (lshamis) [reply mrp/src/mrp/cmd/up.py]: Done

**Review by exhaustin — COMMENTED** [review]
Sorry for the late review, I know this PR is already merged, but wanted to bring up several issues / discussion points regarding UX:
- Attaching to a nonexistent process X simply prints out 'X'. Maybe a more descriptive error message would prevent users from getting confused?
- Currently the attach feature is minimal but great for interacting with processes through stdin. However, unlike attaching to screen or tmux, it doesn't show the history of stdout. Are you planning on adding this feature to enable easier inspecting of processes? 
- Attaching to a process feels non-responsive as nothing is being printed out, which makes it seem like the process is hanging. It would be nice to either print out a message indicating that the shell has attached successfully to some process, and preferably show the prompt after attaching (ex: `>>>` for the python processes in example 15), which currently doesn't show because it's technically being printed to stdout before attaching. 

Also, why was the fixes on `down` reverted? Was the fix deferred to a later PR or was there something we overlooked?

**Pull request merged**